### PR TITLE
TensorGraph.make_estimator()

### PR DESCRIPTION
**Pull request by peastman**
This is an initial implementation.  There are still lots of features to add, testing, etc.  Does the overall approach look right?

### CODE REVIEW

**Comment by rbharath** [comment]
@peastman I like this approach. It looks general, and once we add in documentation, should make it easy for users to convert their TensorGraphs into `tf.Estimator` objects with little trouble.

**Comment by peastman** [comment]
@lilleswing I could use your advice on adding Tensorboard support.  As I understand it, TensorGraph handles this by letting you call `set_summary()` on any layer, which will cause the output of that layer to be logged.  Except there's also a `tb_input` field that can be used to have it log something else instead?  I can't find any examples of how that's used, and it isn't clear to me whether it's even possible to make that work in this context, since we're creating completely different tensors in a different graph.

I've also been looking at Tensorboard documentation, and most examples I find use it for a different purpose.  Instead of logging internal layers in the calculation, they usually log metrics like accuracy or AUC to track how those change during training.  Does Tensorgraph support doing that?

**Comment by peastman** [comment]
Here's another complication I've run into.  Some of our models do strange things with their inputs that abuse the normal API.  For example, MultiTaskClassifier expects its labels to be a 2D tensor, where each element is the index of a class.  It then overrides `default_generator()` to convert this to a 3D tensor with one-hot encoded classes, and that's what actually gets fed into the graph.  This breaks down when we generate an Estimator from the graph, since now the labels are encoded in the wrong way.

I can think of a few approaches to working around this.

1. Just accept that sometimes the Estimator will behave differently from the original model.  For example, that it will want the labels to be encoded in a different way.  This will lead to a *lot* of confusion.

2. Change our models to not do strange things like this.  This might be a good thing in the long run.  For example, I think it's rather confusing that MultiTaskClassifier expects its labels to use one encoding, but produces outputs with a different encoding.  But this would break backward compatibility, and might involve a lot of work.

3. Add a method to TensorGraph that preprocesses all the inputs when generating an Estimator.  The default implementation would do nothing.  Subclasses could override it as necessary.  This would probably be the most straightforward solution, and would work for this case.  But I can't guarantee there aren't other models that do other strange things this wouldn't work for.

**Comment by peastman** [comment]
It looks like we have 16(!) different classes that override `default_generator()`.  Some of them do really complicated things in it.  For example, take a look at the ones in graph_models.py.  If we go with option 3 above (a method for preprocessing inputs), every one of those classes will need to implement it to do the same processing with Tensorflow operations.  In fact, the default implementation should probably check that `default_generator()` has not been overridden, and throw an exception if it has.  Any time you override one, you also need to override the other.

**Comment by lilleswing** [comment]
@peastman I would declare the tb_input field as deprecated as I don't think it has been used.

I would call the major use case for tensorboard integration would be viewing the model with layers as boxes.  That is really where it has been used up until now.  I think in theory tensorboard could support doing things like AUC if you made a layer which calculated it and then turn on logging for that layer.  But again I do not think this is the current use case for tensorboard in TensorGraph.


> Change our models to not do strange things like this. This might be a good thing in the long run. For example, I think it's rather confusing that MultiTaskClassifier expects its labels to use one encoding, but produces outputs with a different encoding. But this would break backward compatibility, and might involve a lot of work.

SparseSoftmaxCrossEntropyWithLogits is actually the preferred way of doing this (not the way we are currently doing this).  I believe this change in the outputs of our models would be relatively straightforward actually.  (Changing inputs to our models is a different story)

> It looks like we have 16(!) different classes that override default_generator(). Some of them do really complicated things in it. For example, take a look at the ones in graph_models.py. If we go with option 3 above (a method for preprocessing inputs), every one of those classes will need to implement it to do the same processing with Tensorflow operations. In fact, the default implementation should probably check that default_generator() has not been overridden, and throw an exception if it has. Any time you override one, you also need to override the other.

I haven't found a good general solution to overriding default_generator.  It tightly couples a model's implementation with data storage but I think we need that specificity.  I do like the idea of throwing an exception if default_generator() has been overridden.

**Comment by peastman** [comment]
> I would declare the tb_input field as deprecated as I don't think it has been used.

Shall I just remove it?  If it's never been used, and if it doesn't work with Estimators, that seems like the safest approach.

> But again I do not think this is the current use case for tensorboard in TensorGraph.

All the Estimator documentation assumes that's how you're using Tensorboard.  For example, see https://www.tensorflow.org/get_started/custom_estimators#tensorboard.

> I haven't found a good general solution to overriding default_generator.

It feels a bit like we're mixing up the featurizer with the model.  A simple principle would be to say that models have features, outputs, and labels.  It's the job of the featurizer to convert your data into whatever input format the model expects, and the outputs and labels should have exactly the same shape and format so they can be directly compared.

Anyway, I'll see what I can do to make the simple cases like MultiTaskClassifier and MultiTaskRegressor work.  For everything else, the authors of the models will need to add support before they can be converted to Estimators.

**Comment by peastman** [comment]
> the outputs and labels should have exactly the same shape and format so they can be directly compared.

To give an example of why this is important, all of the metrics defined in tf.metrics assume this is true.  So if a model doesn't satisfy this requirement, users will be forced to write complicated adapter code just to evaluate metrics.

**Comment by lilleswing** [comment]
> I would declare the tb_input field as deprecated as I don't think it has been used.
>> Shall I just remove it? If it's never been used, and if it doesn't work with Estimators, that seems like the safest approach.

This sounds good to me.

**Comment by rbharath** [comment]
> > I haven't found a good general solution to overriding default_generator.

> It feels a bit like we're mixing up the featurizer with the model. A simple principle would be to say that models have features, outputs, and labels. It's the job of the featurizer to convert your data into whatever input format the model expects, and the outputs and labels should have exactly the same shape and format so they can be directly compared.

I like this principle in general, but worry that changing the existing featurizers could be a large backwards compatibility breaking change. Perhaps we could make some generic wrapper that converts existing featurizers to behave as needed.

**Comment by peastman** [comment]
This now works with MultiTaskClassifier and MultiTaskRegressor.  Just about all of our standard models will also need modification to work with it, but it's best if the authors of those models do the changes.  You need to override `create_estimator_inputs()` and have it generate the correct input tensors for the model.

In some cases, that may not be possible.  I looked into converting MultiTaskFitTransformRegressor, but it involves calls to transformers as part of generating the inputs.  To reproduce the same calculations with Tensorflow operations on tensors, we would need to add that capability to every transformer class.

I made a couple of other changes to enable these classes to work.  The output from MultiTaskClassifier is now probabilities, not logits.  I think that's what people would generally expect them to be in any case.  I also made TensorGraph more tolerant about the shapes of inputs.  It will automatically add or remove trailing dimensions of size 1 to make the input arrays match the corresponding input layers.  This eliminated the need for MultiTaskRegressor to override `default_generator()`.

**Comment by peastman** [comment]
I think this PR is ready, as far as it goes.  We still need to modify all the other TensorGraph subclasses so `make_estimator()` will work with them.  Once this is merged, I'll open an issue to coordinate that.

**Comment by peastman** [comment]
Actually hold off on merging.  I found a problem I need to fix.

**Comment by peastman** [comment]
Ok.  It turned out there wasn't really a problem, but I improved the test cases a bit to be more useful tests.

**Comment by lilleswing** [comment]
I'm going to merge this now.  Have you made the ticket to split up the work for the remaining models?  I can probably do most of it -- but we should have a list.

**Comment by peastman** [comment]
Thanks!  I'll create the issue for that shortly.
